### DISCUSSION TITLE
Should `LowRankOperator` have a rank-truncation method?

### DISCUSSION THREAD

**Discussion by drittelhacker** [pymor/pymor#2285] [Ideas]
The operator is constructed from the left and right VectorArrays L, R and a core NumPy array C. Consider that L, R have themselves been created as concatenations of VectorArrays and C is the block diagonal NumPy array. The result may very well have (numerical) rank-deficiency, even if the originals were fine. This situation could, e.g., happen for a linear combination of two LowRankOperators. In many situations, it would then be desirable to have a more compact representation of the LowRankOperator, that matches the (numerical) rank.

The easiest truncation procedure would 

- compute QR decompositions of L and R,
- form the product of their R factors with C, 
- decide the rank by a truncated SVD of that product,
- update the Q factors of L and R with the unitary factors from the SVD. 
 
Obviously, a truncation tolerance on the SVD would then be required.

**Comment by sdrave** [comment]
Sounds fine to me, as long as the truncation does not happen upon instantiation. Due to immutability, the method would return a new operator.

**Reply by drittelhacker** [reply]
I would aim for something like 
`A = (B  - C).truncate(1e-16)`
so, `B-C` would be temporary, and the extra memory is hardly avoidable with the QRs anyway.

**Comment by sdrave** [comment]
Sounds good.
